### DISCUSSION TITLE
Allow safe allocation of reference types on stack (existing code is unchanged)

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#5542] [Language Ideas] [answered]
We would use stackalloc keyword similiar to in/ref/out keywords, heap keyword similiar to ref new in c++/cli and stackalloc like ref struct.

`stackalloc MyClass stackMyClass = new();`

This will allocate stackMyClass on the stack instead of the heap.
The stackalloc modifier would also make the variable not convertible to non-stackalloc variables:

```
MyClass heapMyClass = new();
stackalloc MyClass stackMyClass = new();
heapMyClass = stackMyClass; // Error: cannot assign stackalloc variable to non-stackalloc variable 
stackMyClass = heapMyClass; // Valid: value of stackMyClass is now on heap (Now the stackalloc modifier is temponarly "removed" from stackMyClass)
heapMyClass = stackMyClass; // Valid: stackMyClass is on to heap
stackMyClass = new(); // New instance allocated on stack (stackalloc once again)
stackMyClass = heap new(); // New instance is on heap (stackalloc temponarly removed)
```

This keyword would also be valid in parameter:

```
void MyMethodA(stackalloc MyClass param) {}
void MyMethodB(MyClass param) {}
stackalloc MyClass myClass = new(); // Allocated on stack
MyMethodA(myClass);
MyMethodA(new()); // Allocated on stack
MyMethodA(heap new()); // Allocated on heap 
MyMethodB(myClass); // Error: stackalloc variable cannot be assigned to non-stackalloc variable
MyMethodB(new()); // Allocated on heap
MyMethodB(heap new()); // Allocated on heap
```

Type modifiers:
We want some types to be only allocable on heap, to do that we will use stackalloc keyword
```
class MyClassA {} // Can only be allocated on heap
stackalloc class MyClassB {} // Can be allocated on heap or stack 'this' is stackalloc
stackalloc struct MyStruct {} // Can only be allocated on stack (works the same as ref struct)
// stackalloc type modifier can be applied to any reference type (including interface)
stackalloc interface IMyInterface {} // Valid
stackalloc IMyInterface myInterface = heap new MyClass(); // Valid: MyClass is on heap (stackalloc temponarly "removed")
myInterface = new MyStruct(); // Valid: on stack (stackalloc once again)
```

Inheritance:
Stackalloc inheritance will work the same as normal but all derived classes from stackalloc class must be stackalloc
```
class MyBase {}
stackalloc class MyClass : MyBase {} // Valid
stackalloc class MyDerivedA : MyClass {} // Valid
class MyDerivedB : MyClass {} // Error: All classes derived from stackalloc class must be stackalloc

stackalloc MyClass myClass = new() // MyClass is on stack
myClass = heap new MyDerivedA() // MyDerivedA is on heap, myClass is temponarly not stackalloc
myClass = new MyDerivedA() // MyDerivedA is on stack, myClass is on stack once again
```

Implementation:
> stackalloc on parameters can be implemented as attribute (like InAttribute/OutAttribute)
> stackalloc on locals can be a compile-time thing
> stackalloc on types can be attributes (like IsByRefLikeAttribute)
> allocating on stack won't need any new instructions it will be just like normal new but instead of calling clr's memory allocation function we will just allocate memory on stack

Pros:
-New code will be a lot faster
-Safe (if i didn't miss anything)
-Doesn't change existing code
-Easy implementation (at least i think)
-Easy syntax (its open for suggestions, may be a little confusing)

**Comment by HaloFour** [comment]
The runtime doesn't currently offer any mechanism to allocate a reference type on the stack.  Any implementation would have to start there.

You'd also need a mechanism to constrain the lifetime of the reference as if it were a `ref` local, all existing code that deals with reference types would be incompatible with these instances as the reference could escape and outlive the instance itself.

**Reply by ghost** [reply]
1. You may be right but, it is actually possible to allocate ref-type on any address and use it
```
byte* memory = stackalloc byte[((int*)typeof(MyClass).TypeHandle.Value)[1]] // Get the size of object on the heap and alloc it on stack
*(IntPtr*)memory = typeof(MyClass).TypeHandle.Value; // Set object type
// (Now the constructor should ge invoked)
MyClass myClass = Unsafe.AsRef<MyClass>(&memory); // myClass is now a valid object
```

2. Im solving that with "stackalloc variable cannot be assigned to non-stackalloc variable", this controls the lifetime of the object, object cannot be assigned to any variable outside the current method like global or array, the object won't be freed until you are 100% done using it.

**Reply by AartBluestoke** [reply]
the problem is escape analysis, and what do you expect to do with MyClass: it is now a *very* dangerous object - even myClass.doSomething() is potentially catastrophic, as doSomething() could say _myStaticList.Add(this);
you can't call functions on myClass, you can't pass myClass into anything that could possibly keep a handle to the memory location (which is every function that doesn't take a ref)
-- Even a simple Logger.Log($"the thing is {myClass}") is potentially dangerous, as string formatters might take and cache (temporarily) a reference, for later lazy evaluation (eg: a log formatters could do this to avoid cpu load on the main thread).

**Reply by HaloFour** [reply]
@Sebaj12 

That's a hack, not an operation supported by the runtime, and it is incredibly dangerous.  I would suggest that the language team would never build a feature that relies on something like that.

But I agree with @AartBluestoke that the bigger problem is that even if you manage to get such an instance it is incredibly dangerous to use and you'd need more language features to try to ensure safe use.  Quite literally no existing code dealing with reference types could be considered safe.

For niche scenarios where you want the performance of stack-allocated data you're better off defining them as `ref struct`.

**Reply by ghost** [reply]
@AartBluestoke I already anwsered that
```
void Log(object obj);
stackalloc MyClass myClass = new();
Log(myClass); // Error: myClass is stackalloc and it cannot be non-stackalloc
stackalloc class MyClass {} // 'this' is stackalloc
```

@HaloFour I agree with you that was a hack, but when you see jit asm that's +- how an object is created its only a proof of concept, that might/will be implemented diffrently but vs 2022 debugger and runtime have no problems with this type of objects

**Reply by HaloFour** [reply]
If these require completely new types to be defined I kind of don't see the point over using `ref struct`.  I don't think there's much point to a type that can be allocated on either the heap or the stack because all code written to work with that type must assume that it's been allocated on the stack.

**Comment by HaloFour** [comment, marked as answer]
The runtime team has been exploring stack allocation of references as an optimization in the JIT:

https://github.com/dotnet/runtime/issues/11192

**Reply by ghost** [reply]
Thank you!
Google never showed this to me.

**Reply by TonyValenti** [reply]
@HaloFour Is this still active work?  It seems as if progress on the work items are no longer happening.

**Reply by HaloFour** [reply]
Probably a better question for the linked discussion.  Last I heard the benefit was too low for the amount of effort necessary but that they might explore more in AOT scenarios, but I have not been keeping tabs.
